File: baekjoon/src/skt/B_2156.java
package skt;

import java.io.*;

public class B_2156 {
	static int[] grape;
	static int[][] dp;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int n = Integer.parseInt(br.readLine());
		grape = new int[n];
		dp = new int[2][n];
		
		for (int i=0;i<n;i++) {
			grape[i] = Integer.parseInt(br.readLine());
			dp[0][i] = dp[1][i] = -1;
		}
		
		System.out.println(Math.max(solution(0, n - 1), solution(0, n - 2)));
	}
	
	static int solution(int cnt, int n) {
		if (n < 0) return 0;
		if (dp[cnt][n] != -1) return dp[cnt][n];
		
		int temp = Math.max(solution(0, n - 2), solution(0, n - 3));
		if (cnt != 1) temp = Math.max(solution(1, n - 1), temp);
		
		return dp[cnt][n] = temp + grape[n];
	}

}

File: src/kotlinsource/dp/B_2156.kt
package kotlinsource.dp

import kotlin.math.max

class B_2156 {
    private var n = 0
    private lateinit var arr: IntArray
    private lateinit var dp: Array<IntArray>

    operator fun invoke() = with(System.`in`.bufferedReader()) {
        n = readLine().toInt()
        arr = IntArray(n)
        dp = Array(n) { IntArray(2) { -1 } }

        for (idx in 0 until n) {
            arr[idx] = readLine().toInt()
        }

        print(max(solution(0, 0), solution(1, 0)))
    }

    private fun solution(idx: Int, isContinue: Int): Int {
        if (idx >= n) return 0
        if (dp[idx][isContinue] != -1) return dp[idx][isContinue]

        var temp = max(solution(idx + 2, 0), solution(idx + 3, 0))
        if (isContinue == 0) {
            temp = max(temp, solution(idx + 1, 1))
        }

        dp[idx][isContinue] = temp + arr[idx]
        return dp[idx][isContinue]
    }
}

fun main() {
    B_2156()()
}